app/recipes/tests/test_recipe_api.py

```diff
@@ -0,0 +1,155 @@
+
+    def test_adding_an_invalid_recipe(self):
+        """Test adding an invalid recipe"""
+
+        payload = {
+            'name': 'Recipe name',
+        }
+
+        res = self.client.post(RECIPES_URL, payload)
+        self.assertEqual(res.status_code, status.HTTP_400_BAD_REQUEST)
+
+    def test_adding_a_recipe(self):
+        """Test adding a recipe successfully"""
+
+        payload = {
+            "name": "Recipe name",
+            "description": "recipe description",
+            "ingredients": [{
+                'name': 'ingredient1'
+            }],
+        }
+
+        res = self.client.post(RECIPES_URL, payload, format='json')
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+
+    def test_adding_a_recipe(self):
+        """Test adding a recipe successfully"""
+
+        payload = {
+            "name": "Recipe name",
+            "description": "recipe description",
+            "ingredients": [{
+                'name': 'ingredient1'
+            }],
+        }
+
+        res = self.client.post(RECIPES_URL, payload, format='json')
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+
+    def test_retrieving_a_detail_recipe(self):
+        """Test retrieving a recipe successfully"""
+
+        recipe = Recipe.objects.create(name='Recipe name', description='description recipe')
+
+        res = self.client.get(get_recipe_detail_url(recipe.id))
+        serializer = RecipeSerializer(res.data)
+
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(res.data['name'], serializer.data['name'])
+        self.assertEqual(res.data['description'], serializer.data['description'])
+        self.assertEqual(len(res.data['ingredients']), len(serializer.data['ingredients']))
+
+    def test_deleting_a_recipe(self):
+        """Test deleting a recipe successfully"""
+
+        recipe = Recipe.objects.create(name='Recipe to be delete', description='description recipe')
+        Ingredient.objects.create(name='Ingredient1', recipe=recipe)
+
+        res = self.client.delete(get_recipe_detail_url(recipe.id))
+
+        self.assertEqual(res.status_code, status.HTTP_204_NO_CONTENT)
+        exists = Recipe.objects.filter(id=recipe.id).exists()
+        self.assertEqual(exists, False)
+        self.assertEqual(len(Ingredient.objects.all()), 0)
+
+    def test_partial_update_recipe(self):
+        """Test partial update for a recipe"""
+
+        recipe = Recipe.objects.create(name='Recipe to be updated', description='description recipe')
+        Ingredient.objects.create(name='Ingredient1', recipe=recipe)
+
+        patch_url = get_recipe_detail_url(recipe.id)
+
+        payload = {
+            'name': 'Recipe updated'
+        }
+
+        res = self.client.patch(patch_url, payload)
+        recipe.refresh_from_db()
+
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(res.data['name'], recipe.name)
```

Comment:
Should `res.data['name']` be equal to `payload['name']`? It looks like the recipe was not updated. Also assert that the record in the DB changed.

Reply:
I see you are asserting against the DB, sorry :man_facepalming: